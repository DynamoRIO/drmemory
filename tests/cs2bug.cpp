/* **********************************************************
 * Copyright (c) 2011-2017 Google, Inc.  All rights reserved.
 * Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
 * **********************************************************/

/* Dr. Memory: the memory debugger
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation;
 * version 2.1 of the License, and no later version.

 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Library General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <iostream>
#include "stdlib.h"

/* our delete mismatches can crash so we use fault handling for a more robust test */
#ifdef UNIX
# include <unistd.h>
# include <signal.h>
# ifdef MACOS
#  define _XOPEN_SOURCE 700 /* required to get POSIX, etc. defines out of ucontext.h */
#  define __need_struct_ucontext64 /* seems to be missing from Mac headers */
# endif
# include <ucontext.h>
# include <errno.h>
# include <assert.h>
/* just use single-arg handlers */
typedef void (*handler_t)(int);
typedef void (*handler_3_t)(int, siginfo_t *, void *);
#endif

#include <setjmp.h>
jmp_buf mark;

#ifdef UNIX
static void
signal_handler(int sig)
{
    if (sig == SIGSEGV)
        longjmp(mark, 1);
    else
        exit(1);
}
static void
intercept_signal(int sig, handler_t handler)
{
    int rc;
    struct sigaction act;
    act.sa_sigaction = (handler_3_t) handler;
    rc = sigemptyset(&act.sa_mask); /* block no signals within handler */
    assert(rc == 0);
    act.sa_flags = SA_NODEFER | SA_SIGINFO | SA_ONSTACK;
    rc = sigaction(sig, &act, NULL);
    assert(rc == 0);
}
#else
/* sort of a hack to avoid the MessageBox of the unhandled exception spoiling
 * our batch runs
 */
# include <windows.h>
/* top-level exception handler */
static LONG
our_top_handler(struct _EXCEPTION_POINTERS * pExceptionInfo)
{
    if (pExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_ACCESS_VIOLATION) {
        longjmp(mark, 1);
    }
    return EXCEPTION_EXECUTE_HANDLER; /* => global unwind and silent death */
}
#endif

static void
test_basic()
{
    /* uninit error */
    int *p = new int;
    if (*p != 10)
        std::cout << "hi" << std::endl;

    /* unaddr error */
    int *a = new int[3];
    /* on xp64x2cpu vm, w/ VS2005 SP1, heap assert fires: detects
     * the overflow if do a[3], so doing a[4], which is not detected.
     */
    a[4] = a[4];
    if (setjmp(mark) == 0)
        delete a; /* also a mismatch */

    /* zero-sized (i#1145) and "prev malloc" reports */
    a = new int[0];
    *((char *)a + 0) = 4;
    delete [] a;

    a = new int[0];
    *((char *)a + 1) = 4;
    delete [] a;

    a = new int[8];
    *((char *)a + 8*sizeof(int)) = 4;
    delete [] a;
}

class hasdtr {
public:
    hasdtr() { x = new int[7]; }
    ~hasdtr() { delete[] x; }
    int *x;
    int y;
    char z;
};

class parA {
public:
    parA() { a = "parA"; };
    virtual ~parA() {}
    virtual const char *getval() const { return a; }
    const char *a;
};

class parB {
public:
    parB() { b = "parB"; }
    virtual ~parB() {}
    virtual const char *getval() const { return b; }
    virtual const char *myfunc() const { return b; }
    const char *b;
};

class childAB : public parA, public parB {
public:
    childAB() { ab = "childAB"; }
    virtual ~childAB() {}
    virtual const char *getval() const { return ab; }
    const char *ab;
};

static void
test_leaks()
{
    /* test mid-chunk std::string leak (PR 535344) */
    static std::string *str = new std::string("leak");
    std::cout << "size=" << str->size() <<
        " capacity=" << str->capacity() << std::endl;

    /* test mid-chunk pointer in leak due to new[] header (PR 484544)
     * (header is only present if class has destructor)
     */
    static hasdtr *leak = new hasdtr[4];
    leak[0].y = 0;

    /* test mid-chunk pointer in leak due to multiple inheritance by
     * casting to the 2nd of the two parent classes (PR 484544)
     */
    static parB *multi = (parB *) new childAB();
    std::cout << "getval: " << multi->getval() << std::endl;
    std::cout << "myfunc: " << multi->myfunc() << std::endl;

    /* test PR 576032 (dependent leaks): std::string shouldn't show up */
#ifndef WINDOWS
    /* FIXME PR 587093: disabling on Windows until figure out why callstack messed up */
    std::string onstack = "leakme";
    static std::string *outer = new std::string(onstack);
    outer = NULL;
#endif
}

static void
test_exception()
{
    try {
        std::cout << "throwing exception" << std::endl;
        throw std::exception();
    } catch (std::exception&) {
        std::cout << "caught exception" << std::endl;
    }
}

static void
test_mismatch_dtr()
{
#ifndef SKIP_MISMATCH_DTR
    /* /MTd, we skip the destructor mismatches, as they end up raising
     * heap assertions that we can't recover from
     */
    hasdtr *x = new hasdtr[7];
    if (setjmp(mark) == 0)
        delete x;
    x = new hasdtr[7];
    if (setjmp(mark) == 0)
        free(x);
    x = (hasdtr *) malloc(42); /* big enough for hasdtr but same size for x64 */
    if (setjmp(mark) == 0)
        delete x;
    x = (hasdtr *) malloc(42); /* big enough for hasdtr but same size for x64 */
    if (setjmp(mark) == 0)
        delete[] x; /* unaddr reading size + dtr calls might crash before mismatch */
    /* not a mismatch, but test debug operator del (i#500) */
    x = new hasdtr[7];
    delete[] x;
#endif
}

static void
test_mismatch_int()
{
    int *x = new int[7];
    if (setjmp(mark) == 0)
        delete x; /* technically no adverse effects since no destructor */
    x = new int[7];
    if (setjmp(mark) == 0)
        free(x);
    x = (int *) malloc(7);
    if (setjmp(mark) == 0)
        delete x;
    x = (int *) malloc(7);
    if (setjmp(mark) == 0)
        delete[] x;
    /* not a mismatch, but test debug operator del (i#500) */
    x = new int[7];
    delete[] x;
}

int main()
{
#ifdef UNIX
    intercept_signal(SIGSEGV, signal_handler);
#else
    SetUnhandledExceptionFilter((LPTOP_LEVEL_EXCEPTION_FILTER) our_top_handler);
#endif

    test_leaks();

    test_basic();

    test_exception();

    test_mismatch_dtr();
    test_mismatch_int();

    std::cout << "bye" << std::endl;

    /* mismatches above end up causing RtlpCoalesceFreeBlocks to crash resulting
     * in failing app exit code: simpler to just exit
     */
    exit(0);

    return 0;
}
