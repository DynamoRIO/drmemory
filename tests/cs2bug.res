# **********************************************************
# Copyright (c) 2011-2019 Google, Inc.  All rights reserved.
# Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
: UNINITIALIZED READ
cs2bug.cpp:87
: UNADDRESSABLE ACCESS beyond heap bounds: reading 4 byte(s)
cs2bug.cpp:95
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:97
memory was allocated here:
cs2bug.cpp:91
: UNADDRESSABLE ACCESS beyond heap bounds: writing 1 byte(s)
cs2bug.cpp:101
refers to 0 byte(s) beyond last valid byte in prior malloc
: UNADDRESSABLE ACCESS beyond heap bounds: writing 1 byte(s)
cs2bug.cpp:105
refers to 1 byte(s) beyond last valid byte in prior malloc
: UNADDRESSABLE ACCESS beyond heap bounds: writing 1 byte(s)
cs2bug.cpp:109
refers to 0 byte(s) beyond last valid byte in prior malloc
##################################################
# test_mismatch_dtr()
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:197
memory was allocated here:
cs2bug.cpp:195
%OPTIONAL # Linux
: INVALID HEAP ARGUMENT to free
%ENDOPTIONAL
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with free
cs2bug.cpp:200
memory was allocated here:
cs2bug.cpp:198
%OPTIONAL # only when wrapping
: UNINITIALIZED READ
cs2bug.cpp:116
cs2bug.cpp:203
%ENDOPTIONAL
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete
cs2bug.cpp:203
memory was allocated here:
cs2bug.cpp:201
%OPTIONAL # VS2008 Win7
: UNINITIALIZED READ
cs2bug.cpp:203
%ENDOPTIONAL
: UNADDRESSABLE ACCESS
cs2bug.cpp:206
%OPTIONAL
# MinGW xp64 crashes rather than reporting final mismatch
: UNADDRESSABLE ACCESS
cs2bug.cpp:206
%ENDOPTIONAL
%OPTIONAL
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete[]
cs2bug.cpp:206
memory was allocated here:
cs2bug.cpp:204
%ENDOPTIONAL
%OPTIONAL # Linux
: INVALID HEAP ARGUMENT to free
%ENDOPTIONAL
##################################################
# test_mismatch_int()
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:218
memory was allocated here:
cs2bug.cpp:216
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with free
cs2bug.cpp:221
memory was allocated here:
cs2bug.cpp:219
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete
cs2bug.cpp:224
memory was allocated here:
cs2bug.cpp:222
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete[]
cs2bug.cpp:227
memory was allocated here:
cs2bug.cpp:225
##################################################
# leaks
%OUT_OF_ORDER
: LEAK 4 direct bytes + 0 indirect bytes
cs2bug.cpp:86
%if UNIX
# FIXME PR 587093: string code disabled for now on Windows
%if X32
: LEAK 4 direct bytes + 19 indirect bytes
%endif
%if X64
: LEAK 8 direct bytes + 31 indirect bytes
%endif
cs2bug.cpp:172
# Nested %if is only supported with "endif UNIX".
%endif UNIX
%OPTIONAL # Linux/VS2005
%if X32
: LEAK 88 direct bytes + 168 indirect bytes
%endif
%if X64
: LEAK 120 direct bytes + 168 indirect bytes
%endif
cs2bug.cpp:195
%ENDOPTIONAL
%if X32
: LEAK 88 direct bytes + 196 indirect bytes
%endif
%if X64
: LEAK 120 direct bytes + 196 indirect bytes
%endif
cs2bug.cpp:198
: LEAK 42 direct bytes + 0 indirect bytes
cs2bug.cpp:204
