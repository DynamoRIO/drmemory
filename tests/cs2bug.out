# **********************************************************
# Copyright (c) 2011-2020 Google, Inc.  All rights reserved.
# Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
%ANYLINE
# MSVC
size=4 capacity=15
# Linux, MinGW
size=4 capacity=4
%ENDANYLINE
getval: childAB
myfunc: parB
hi
bye
~~Dr.M~~ ERRORS FOUND:
%ANYLINE
# MinGW xp64
~~Dr.M~~       6 unique,     6 total unaddressable access(es)
# Linux wrap
~~Dr.M~~       4 unique,     4 total unaddressable access(es)
# everything else
~~Dr.M~~       5 unique,     5 total unaddressable access(es)
%ENDANYLINE
%ANYLINE
# Linux/MinGW/VS2005
~~Dr.M~~       2 unique,     2 total uninitialized access(es)
# VS2008 Win7
~~Dr.M~~       3 unique,     3 total uninitialized access(es)
%ENDANYLINE
%ANYLINE
# Linux/MinGW
~~Dr.M~~      11 unique,    12 total invalid heap argument(s)
# MinGW xp64
~~Dr.M~~      10 unique,    11 total invalid heap argument(s)
# MSVC
~~Dr.M~~       9 unique,    12 total invalid heap argument(s)
%ENDANYLINE
~~Dr.M~~       0 unique,     0 total warning(s)
%ANYLINE
# Windows: FIXME PR 587093: string code disabled for now
# MSVC: mismatches don't end up freeing anything
~~Dr.M~~       4 unique,     4 total,   %ANY% byte(s) of leak(s)
# Linux/MinGW.  Total count varies by compiler version.
~~Dr.M~~       5 unique,     5 total,   %ANY% byte(s) of leak(s)
%ENDANYLINE
%ANYLINE
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
# Recent Linux g++ ends up with a possible.
~~Dr.M~~       1 unique,     1 total,   %ANY% byte(s) of possible leak(s)
%ENDANYLINE
